File: FSharp/BookingApi/MaîtreD.fs
module Ploeh.Samples.MaîtreD

open System

type Reservation = {
    Date : DateTimeOffset
    Quantity : int
    IsAccepted : bool
}

// int -> Reservation list -> Reservation -> Reservation option
let tryAccept capacity reservations reservation =
    let reservedSeats =
        reservations |> List.sumBy (fun x -> x.Quantity)
    if reservedSeats + reservation.Quantity <= capacity
    then { reservation with IsAccepted = true } |> Some
    else None

File: Haskell/MaîtreD.hs
module MaîtreD where

import Data.Time (ZonedTime (zonedTimeToLocalTime, zonedTimeZone))

data Reservation =
  Reservation { date :: ZonedTime, quantity :: Int, isAccepted :: Bool }
  deriving (Show)

instance Eq Reservation where
  x == y =
       zonedTimeZone        (date x) == zonedTimeZone        (date y)
    && zonedTimeToLocalTime (date x) == zonedTimeToLocalTime (date y)
    && quantity                   x  ==  quantity                  y
    && isAccepted                 x  == isAccepted                 y


tryAccept :: Int -> [Reservation] -> Reservation -> Maybe Reservation
tryAccept capacity reservations reservation =
  let reservedSeats = sum $ map quantity reservations
  in  if reservedSeats + quantity reservation <= capacity
      then Just $ reservation { isAccepted = True }
      else Nothing
